File: src/Milekic.YoLo/VersionedList.fs
namespace Milekic.YoLo

open System.Collections.Generic
open System.Collections.Immutable

module VersionedList =
    type Change<'T> =
        | Add of 'T
        | AddRange of ImmutableArray<'T>
        | Remove of 'T
        | RemoveAt of int
        | Insert of int * 'T
        | InsertRange of int * ImmutableArray<'T>
        | SetItem of int * 'T
        | Replace of 'T * 'T
    let makeChangeBuilder (list : ImmutableList<_>) =
        let builder = list.ToBuilder()
        let folder =
            function
            | Add item -> builder.Add item
            | AddRange items -> builder.AddRange items
            | Remove item -> builder.Remove item |> ignore
            | RemoveAt index -> builder.RemoveAt index
            | Insert (index, item) -> builder.Insert(index, item)
            | InsertRange (index, items) -> builder.InsertRange(index, items)
            | SetItem (index, item) -> builder[index] <- item
            | Replace (oldItem, newItem) ->
                let index = builder.IndexOf oldItem
                if index >= 0 then builder[index] <- newItem
        let runner = fun () -> builder.ToImmutable()
        folder, runner

open VersionedList

type VersionedList<'T when 'T : equality> private (store, history, hashCodes) =
    new() = VersionedList(ImmutableList.Empty, ImmutableList.Empty,  ImmutableList.Empty)

    member _.Store = store
    member _.History = history
    member _.Version = history.Count
    member private _.State = (store, history, hashCodes)

    member this.GetChangesSince(other : VersionedList<'T>) =
        let firstNewChange = other.Version + 1
        if firstNewChange >= this.Version then None else

        if (hashCodes[firstNewChange] <> other.GetHashCode())
        then None
        else let x = history |> Seq.toArray in Some x[firstNewChange..]

    override this.Equals(obj) =
        match obj with
        | null -> false
        | _ when LanguagePrimitives.PhysicalEquality (box this) obj -> true
        | _ when obj.GetType() <> this.GetType() -> false
        | _ -> this.Equals(obj :?> VersionedList<'T>)
    member this.Equals(other : VersionedList<'T>) = this.State = other.State
    override this.GetHashCode() = this.State.GetHashCode()

    member _.ClearHistory () =
        VersionedList(store, ImmutableList.Empty, ImmutableList.Empty)
    member this.Add(item) =
        VersionedList(
            store.Add(item),
            history.Add(Add item),
            hashCodes.Add(this.GetHashCode()))
    member this.AddRange(items : 'T seq) =
        let items = items.ToImmutableArray()
        VersionedList(
            store.AddRange(items),
            history.Add(AddRange items),
            hashCodes.Add(this.GetHashCode()))
    member this.Insert(index, item) =
        VersionedList(
            store.Insert(index, item),
            history.Add(Insert(index, item)),
            hashCodes.Add(this.GetHashCode()))
    member this.InsertRange(index, items : 'T seq) =
        let items = items.ToImmutableArray()
        VersionedList(
            store.InsertRange(index, items),
            history.Add(InsertRange(index, items)),
            hashCodes.Add(this.GetHashCode()))
    member this.Remove(item) =
        VersionedList(
            store.Remove item,
            history.Add(Remove item),
            hashCodes.Add(this.GetHashCode()))
    member this.RemoveAt(index) =
        VersionedList(
            store.RemoveAt index,
            history.Add(RemoveAt index),
            hashCodes.Add(this.GetHashCode()))
    member this.SetItem(index, item) =
        VersionedList(
            store.SetItem(index, item),
            history.Add(SetItem (index, item)),
            hashCodes.Add(this.GetHashCode()))
    member this.Replace(oldValue, newValue) =
        VersionedList(
            store.Replace(oldValue, newValue),
            history.Add(Replace (oldValue, newValue)),
            hashCodes.Add(this.GetHashCode()))

    interface IReadOnlyList<'T> with
        member this.GetEnumerator() : IEnumerator<'T> = store.GetEnumerator()
        member this.GetEnumerator() : System.Collections.IEnumerator = store.GetEnumerator()
        member this.Count = store.Count
        member this.Item with get index = store[index]

File: src/Milekic.YoLo.CSharp/VersionedList.cs
using System.Collections;
using System.Collections.Immutable;

namespace Milekic.YoLo;

public record VersionedList<T> : IReadOnlyList<T>
{
    public abstract record Change
    {
        public interface IVisitor
        {
            void Visit(AddRange addRange);
            void Visit(RemoveRange removeRange);
            void Visit(InsertRange insertRange);
            void Visit(SetItem setItem);
            void Visit(ReplaceItem replaceItem);
            void Visit(RemoveAt removeAt);
        }

        private Change() { }

        public abstract void Accept(IVisitor current);

        public record AddRange(ImmutableArray<T> Items) : Change
        {
            public override void Accept(IVisitor visitor) => visitor.Visit(this);
        }
        public record RemoveRange(ImmutableArray<T> Items) : Change
        {
            public override void Accept(IVisitor visitor) => visitor.Visit(this);
        }
        public record RemoveAt(int Index) : Change
        {
            public override void Accept(IVisitor visitor) => visitor.Visit(this);
        }
        public record InsertRange(int Index, ImmutableArray<T> Items) : Change
        {
            public override void Accept(IVisitor visitor) => visitor.Visit(this);
        }
        public record SetItem(int Index, T Item) : Change
        {
            public override void Accept(IVisitor visitor) => visitor.Visit(this);
        }
        public record ReplaceItem(T OldItem, T NewItem) : Change
        {
            public override void Accept(IVisitor visitor) => visitor.Visit(this);
        }
    }

    public class ChangeFolder : Change.IVisitor
    {
        private readonly ImmutableList<T>.Builder Builder;
        public ChangeFolder(ImmutableList<T> source)
        {
            Builder = source.ToBuilder();
        }
        public ImmutableList<T> GetResult() => Builder.ToImmutableList();

        public void Visit(Change.AddRange addRange) => Builder.AddRange(addRange.Items);
        public void Visit(Change.RemoveRange removeRange)
        {
            foreach (var item in removeRange.Items)
            {
                Builder.Remove(item);
            }
        }
        public void Visit(Change.InsertRange insertRange) =>
            Builder.InsertRange(insertRange.Index, insertRange.Items);
        public void Visit(Change.SetItem setItem) => Builder[setItem.Index] = setItem.Item;
        public void Visit(Change.ReplaceItem replaceItem) =>
            Builder[Builder.IndexOf(replaceItem.OldItem)] = replaceItem.NewItem;
        public void Visit(Change.RemoveAt removeAt) => Builder.RemoveAt(removeAt.Index);
    }

    public readonly ImmutableList<T> Store;
    public readonly ImmutableList<Change> History;
    private readonly ImmutableList<int> HashCodes;

    public int Version => History.Count;
    public int Count => Store.Count;
    public T this[int index] => Store[index];

    public VersionedList() : this(
        ImmutableList<T>.Empty,
        ImmutableList<Change>.Empty,
        ImmutableList<int>.Empty) { }
    private VersionedList(ImmutableList<T> store, ImmutableList<Change> history, ImmutableList<int> hashCodes)
    {
        Store = store;
        History = history;
        HashCodes = hashCodes;
    }

    public IEnumerable<Change> GetChangesSince(VersionedList<T> old)
    {
        var firstNewChange = old.Version + 1;
        if (firstNewChange >= Version || HashCodes[firstNewChange] != old.GetHashCode())
        {
            throw new ArgumentException("Current version is not a newer version of old");
        }
        return History.Skip(old.Version);
    }

    public IEnumerator<T> GetEnumerator() => Store.GetEnumerator();
    IEnumerator IEnumerable.GetEnumerator() => ((IEnumerable)Store).GetEnumerator();

    public VersionedList<T> Add(T item) => new(
        Store.Add(item),
        History.Add(new Change.AddRange(ImmutableArray.Create(item))),
        HashCodes.Add(GetHashCode()));
    public VersionedList<T> AddRange(IEnumerable<T> items)
    {
        var array = items.ToImmutableArray();
        return new(
            Store.AddRange(array),
            History.Add(new Change.AddRange(array)),
            HashCodes.Add(GetHashCode()));
    }
    public VersionedList<T> Insert(int index, T item) => new(
        Store.Insert(index, item),
        History.Add(new Change.InsertRange(index, ImmutableArray.Create(item))),
        HashCodes.Add(GetHashCode()));
    public VersionedList<T> InsertRange(int index, IEnumerable<T> items)
    {
        var array = items.ToImmutableArray();
        return new(
            Store.InsertRange(index, array),
            History.Add(new Change.InsertRange(index, array)),
            HashCodes.Add(GetHashCode()));
    }
    public VersionedList<T> Remove(T item) => new(
        Store.Remove(item),
        History.Add(new Change.RemoveRange(ImmutableArray.Create(item))),
        HashCodes.Add(GetHashCode()));
    public VersionedList<T> RemoveRange(IEnumerable<T> items)
    {
        var array = items.ToImmutableArray();
        return new(
            Store.RemoveRange(array),
            History.Add(new Change.RemoveRange(array)),
            HashCodes.Add(GetHashCode()));
    }
    public VersionedList<T> RemoveAt(int index) => new(
        Store.RemoveAt(index),
        History.Add(new Change.RemoveAt(index)),
        HashCodes.Add(GetHashCode()));
    public VersionedList<T> SetItem(int index, T value) => new(
        Store.SetItem(index, value),
        History.Add(new Change.SetItem(index, value)),
        HashCodes.Add(GetHashCode()));
    public VersionedList<T> Replace(T oldItem, T newItem) => new(
        Store.Replace(oldItem, newItem),
        History.Add(new Change.ReplaceItem(oldItem, newItem)),
        HashCodes.Add(GetHashCode()));
}
